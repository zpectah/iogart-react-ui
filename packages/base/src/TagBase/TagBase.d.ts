import { TagBaseProps } from './types';

declare const TagBase: TagBaseProps;

export default TagBase;
