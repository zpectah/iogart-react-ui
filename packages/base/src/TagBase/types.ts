import {
    MouseEvent,
    MouseEventHandler,
    ReactNode,
} from 'react';

import {
    IogartButtonElement,
    IogartCommonUiProps,
    IogartSpanElement,
} from '@iogart-react-ui/types';
import { TagBaseElementKeys } from './enums';

export interface TagBaseProps<T = IogartSpanElement | IogartButtonElement> extends IogartCommonUiProps {
    label?: ReactNode;
    component?: keyof typeof TagBaseElementKeys;
    clickable?: boolean;
    onClick?: MouseEventHandler<T> | undefined;
    onDoubleClick?: (event: MouseEvent<T, MouseEvent>) => void;
    onClickDelete?: MouseEventHandler<IogartButtonElement> | undefined;
    deleteIcon?: ReactNode;
    onMouseEnter?: MouseEventHandler;
    onMouseLeave?: MouseEventHandler;
}

export interface useTagBaseParameters extends TagBaseProps {}
export interface useTagBaseReturn extends useTagBaseParameters {}
