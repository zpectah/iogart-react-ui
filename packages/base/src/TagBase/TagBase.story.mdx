import { Meta, Story } from '@storybook/addon-docs';

import TagBase from './TagBase';
export const Template = (args) => <TagBase {...args} />;

<Meta
    title="Components/Status/Tag"
    component={TagBase}
/>

# Button
`@iogart-react-ui/base/TagBase`

```javascript
import { TagBase } from 'iogart-react-ui/base';
```

<Story
  name="Base"
  args={{
    label: 'Tag name',
  }}>
  {Template.bind({})}
</Story>
