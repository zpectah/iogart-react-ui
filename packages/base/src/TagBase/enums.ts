export enum TagBaseElementKeys {
    span = 'span',
    div = 'div',
}
