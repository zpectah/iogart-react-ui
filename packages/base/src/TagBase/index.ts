export { default as TagBase } from './TagBase';
export { default as useTagBase } from './useTagBase';
export * from './types';
export * from './enums';
