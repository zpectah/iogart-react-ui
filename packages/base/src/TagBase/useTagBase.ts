import { useState } from 'react';

import { Sx } from '@iogart-react-ui/core';
import { getElementClassName } from '@iogart-react-ui/utils';
import {
    IogartAnchorElement,
    IogartButtonElement,
    IogartClickEvent,
    IogartDoubleClickEvent,
    IogartMouseEvent,
} from '@iogart-react-ui/types';
import { useTagBaseParameters, useTagBaseReturn } from './types';

const useTagBase = (props: useTagBaseParameters) => {
    const {
        className,
        onClick,
        onDoubleClick,
        onMouseEnter,
        onMouseLeave,
        ...rest
    } = props;

    const [ isHover, setHover ] = useState(false);

    const updatedClassName = getElementClassName(
        className,
        [ 'iogart-tag-base' ],
        { hover: isHover },
        { ui: true, base: true },
    );

    const clickHandler = (event: IogartClickEvent<IogartAnchorElement | IogartButtonElement>) => {
        if (onClick) onClick(event);
    };
    const doubleClickHandler = (event: IogartDoubleClickEvent<IogartAnchorElement | IogartButtonElement>) => {
        if (onDoubleClick) onDoubleClick(event);
    };
    const mouseEnterHandler = (event: IogartMouseEvent) => {
        setHover(true);
        if (onMouseEnter) onMouseEnter(event);
    };
    const mouseLeaveHandler = (event: IogartMouseEvent) => {
        setHover(false);
        if (onMouseLeave) onMouseLeave(event);
    };

    const eventsProps = {
        onClick: clickHandler,
        onDoubleClick: doubleClickHandler,
        onMouseEnter: mouseEnterHandler,
        onMouseLeave: mouseLeaveHandler,
    };

    return {
        className: updatedClassName,
        ...eventsProps,
        ...rest
    } as useTagBaseReturn;
};

export default useTagBase;
