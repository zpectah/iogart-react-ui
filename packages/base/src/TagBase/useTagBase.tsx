import { ReactNode } from 'react';

import { useTagBaseParameters, useTagBaseReturn } from './types';
import { Close } from '@iogart-react-ui/icons/src';

const useTagBase = (props: useTagBaseParameters) => {
    const { className, deleteIcon, ...rest } = props;

    const getClassName = () => {

        return `iogart-base iogart-tag-base${className && ` ${className}`}`;
    };

    const getDeleteIcon = () => {

        return deleteIcon ? deleteIcon : <Close size="1rem" />;
    };

    return {
        className: getClassName(),
        deleteIcon: getDeleteIcon() as ReactNode,
        ...rest,
    } as useTagBaseReturn;
};

export default useTagBase;
