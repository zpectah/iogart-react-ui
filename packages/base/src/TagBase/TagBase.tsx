import React, { FC, createElement, useMemo } from 'react';

import { IogartButtonElement } from '@iogart-react-ui/types';
import { useButtonBase } from '@iogart-react-ui/base';
import { Close } from '@iogart-react-ui/icons';
import { TagBaseElementKeys } from './enums';
import { TagBaseProps } from './types';
import useTagBase from './useTagBase';

const TagBase: FC<TagBaseProps> = (props) => {
    const {
        component = TagBaseElementKeys.span,
        label = '',
        clickable,
        deleteIcon,
        onClickDelete,
        ...rest
    } = props;

    const { ...restOfUpdatedProps } = useTagBase({ clickable, ...rest });

    const buttonProps = useButtonBase({
        className: 'iogart-tag-button',
        onClick: onClickDelete && onClickDelete,
    }) as IogartButtonElement;

    const childrenNode = useMemo(() => {
        let children = label;
        if (clickable && onClickDelete) {
            const closeButton = (
                <button {...buttonProps}>
                    {deleteIcon ? deleteIcon : <Close />}
                </button>
            );
            children = (
                <>
                    {children}
                    {closeButton}
                </>
            );
        }

        return children;
    }, [ label, clickable, onClickDelete, deleteIcon, buttonProps ]);

    return createElement(
        component,
        { ...restOfUpdatedProps },
        childrenNode,
    );
};

export default TagBase;
