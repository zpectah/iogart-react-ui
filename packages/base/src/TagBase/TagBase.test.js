import renderer from 'react-test-renderer';

import TagBase from './TagBase';

it('Tag Base - onMouseEnter & onMouseLeave', () => {

    const component = renderer.create(
        <TagBase label="TagBase" />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
        tree.props?.onMouseEnter();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
       tree.props?.onMouseLeave();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});
