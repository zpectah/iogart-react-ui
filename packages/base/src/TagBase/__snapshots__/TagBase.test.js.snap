// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Base - onMouseEnter & onMouseLeave 1`] = `
<span
  className="iogart-ui iogart-base iogart-tag-base"
  onClick={[Function]}
  onDoubleClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  TagBase
</span>
`;

exports[`Tag Base - onMouseEnter & onMouseLeave 2`] = `
<span
  className="iogart-ui iogart-base iogart-tag-base iogart--hover"
  onClick={[Function]}
  onDoubleClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  TagBase
</span>
`;

exports[`Tag Base - onMouseEnter & onMouseLeave 3`] = `
<span
  className="iogart-ui iogart-base iogart-tag-base"
  onClick={[Function]}
  onDoubleClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  TagBase
</span>
`;
